shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -365,201 +376,239 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+            ), reverseRigid2)
+            val mthTy = td2.wrapMethod(bodyTy)
+            if (rhs.isRight || !declared.isDefinedAt(nme.name)) {
+              if (top) thisCtx.addMth(S(td.nme.name), nme.name, mthTy)
+              thisCtx.addMth(N, nme.name, mthTy)
+            }
+            nme.name -> MethodType(thisCtx.lvl, S(ProvType(bodyTy.body)(prov)), td2.nme)
           }
+          MethodDefs(td2.nme, filterTR(tr.expand).map(rec(_)(thisCtx)),
+            td2.mthDecls.iterator.map(go).toMap, td2.mthDefs.iterator.map(go).toMap)
         }
-      ctx.copy(mthDecls = newMthDecls.toMap, mthDefs = newMthDefs.toMap).tap(checkSubsume(_))
-    }
-    /* Perform subsumption checking on method declarations and definitions
-     * by applying type arguments to method signatures of the parents and rigidifying class type variables. */
-    def checkSubsume(implicit ctx: Ctx): Unit = {
-      newDefs.filter(td => ctx.tyDefs.isDefinedAt(td.nme.name)).foreach { td =>
+        val mds = rec(TypeRef(td.nme, rigidtargs)(prov), true)(ctx)
+        checkSubsume(td, mds)
+      }
+      /* Perform subsumption checking on method declarations and definitions by rigidifying class type variables,
+       * then register the method signitures in the context */
+      def checkSubsume(td: TypeDef, mds: MethodDefs): Unit = {
         implicit val prov: TypeProvenance = tp(td.toLoc, "type definition")
         val tn = td.nme
-        val decls = ctx.mthDecls(tn.name)
-        val defs = ctx.mthDefs(tn.name)
-        val rigidSubsMap = td.targsMap.map { 
-          case s -> (tv: TypeVariable) => tv -> freshenAbove(ctx.lvl, tv, true)
-          case _ => die
-        }.toMap[SimpleType, SimpleType]
-        val subsMap = getSubsMap(tn.name)
-        def ss(mt: PolymorphicType, bmt: TypeScheme, bno: Option[Str] = N)(implicit raise: Raise, prov: TypeProvenance) = bmt match {
-          case pt: PolymorphicType => constrain(subst(mt, rigidSubsMap).instantiate,
-            subst(bno.fold(pt)(bn => subst(pt, subsMap(bn))), rigidSubsMap).rigidify)(raise, prov, ctx)
-          case st: SimpleType => constrain(subst(mt, rigidSubsMap).instantiate,
-            subst(bno.fold(st)(bn => subst(st, subsMap(bn))), rigidSubsMap))(raise, prov, ctx)
+        val MethodDefs(_, _, decls, defns) = mds
+        val MethodDefs(_, _, declsInherit, defnsInherit) = mds.propagate()
+        val rigidtargs = td.targs.map(freshenAbove(ctx.lvl, _, true))
+        val targsMap = td.targs.lazyZip(rigidtargs).toMap[SimpleType, SimpleType]
+        def ss(mt: MethodType, bmt: MethodType)(implicit prov: TypeProvenance) =
+          constrain(subst(mt.toPT, targsMap).instantiate, subst(bmt.toPT, targsMap).rigidify)
+        def registerImplicit(nme: Str, mthTy: MethodType) = ctx.getMth(N, nme) match {
+          case S(MethodType(_, _, parents)) if parents.iterator.map(prt => Var(prt.name.decapitalize)).toSet
+              .subsetOf(ctx.allBaseClassesOf(tn.name)) =>
```

Comment:
This non-trivial condition really needs some doc. Be empathetic to the people who will be trying to understand your intention later on!! 😉
Despite the fact I _know_ the semantics of methods very well, I have no idea what this is checking...
```suggestion
              .subsetOf(ctx.allBaseClassesOf(tn.name)) =>
                // If [...] then we don't need to do anything
```

Reply:
Addressed in c5f6b0817ec9905325a697a9281652782b2d2e30

Reply:
Great, the comments make things infinitely clearer now, especially the example.